File: app/src/main/java/com/akari/ppx/xp/hook/code/purity/ImageHook.java
package com.akari.ppx.xp.hook.code.purity;

import com.akari.ppx.xp.hook.code.SuperbHook;

import java.util.ArrayList;
import java.util.concurrent.FutureTask;

import de.robv.android.xposed.XC_MethodReplacement;

import static com.akari.ppx.common.constant.Prefs.SAVE_IMAGE;
import static de.robv.android.xposed.XposedHelpers.callMethod;

public class ImageHook extends SuperbHook {
	@Override
	protected void onHook(ClassLoader cl) {
		hookMethod(SAVE_IMAGE, "com.sup.android.m_gallery.NewGalleryActivity", "onDownload", new XC_MethodReplacement() {
			private boolean isThumb(Object obj1, Object obj2) {
				return Math.abs((int) callMethod(obj1, "getWidth") - (int) callMethod(obj2, "getWidth")) < 5 && Math.abs((int) callMethod(obj1, "getHeight") - (int) callMethod(obj2, "getHeight")) < 5;
			}

			@Override
			protected Object replaceHookedMethod(MethodHookParam param) throws Throwable {
				Object currentImage = ((ArrayList) callMethod(param.thisObject, "getImages")).get((int) callMethod(callMethod(param.thisObject, "getVpGallery"), "getCurrentItem"));
				ArrayList thumbs = (ArrayList) callMethod(param.thisObject, "getThumbs");
				boolean useCdn = false;
				String url = null;
				if (thumbs == null || !isThumb(currentImage, thumbs.get((int) callMethod(callMethod(param.thisObject, "getVpGallery"), "getCurrentItem")))) {
					url = "https://sf1-nhcdn-tos.pstatp.com/obj/" + callMethod(currentImage, "getUri");
					FutureTask<String> task = new FutureTask<>(new WebTask(1, url, cl));
					new Thread(task).start();
					if (!"image/webp".equals(task.get()))
						useCdn = true;
				}
				ArrayList urlList = (ArrayList) callMethod(currentImage, "getUrlList");
				boolean isGif = (boolean) callMethod(currentImage, "isGif");
				if (!useCdn) {
					url = (String) callMethod(urlList.get(0), "getUrl");
					url = url.substring(0, url.lastIndexOf('.')) + (isGif ? ".gif" : ".png");
				}
				callMethod(urlList.get(0), "setUrl", url);
				callMethod(param.thisObject, "downloadImage", urlList, isGif);
				return null;
			}
		});
	}
}

File: app/src/main/java/com/akari/ppx/xp/hook/purity/ImageHook.kt
@file:Suppress("unused")

package com.akari.ppx.xp.hook.purity

import android.app.Activity
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import com.akari.ppx.utils.*
import com.akari.ppx.xp.Init.cl
import com.akari.ppx.xp.hook.SwitchHook
import kotlin.math.abs

class ImageHook : SwitchHook("save_image") {
    override fun onHook() {
        "com.sup.android.m_gallery.NewGalleryActivity".replaceMethod(cl, "onDownload") { param ->
            CoroutineScope(Dispatchers.Main).launch {
                with(param.thisObject as Activity) {
                    val index = callMethod("getVpGallery")?.callMethodAs<Int>("getCurrentItem")!!
                    val current = callMethodAs<ArrayList<*>>("getImages")[index]
                    val thumbs = callMethodAs<ArrayList<*>?>("getThumbs")
                    var useCdn = false
                    var url = ""

                    fun isThumb(img1: Any, img2: Any) =
                        abs(img1.callMethodAs<Int>("getWidth") - img2.callMethodAs<Int>("getWidth")) < 5
                                && abs(
                            img1.callMethodAs<Int>("getHeight") - img2.callMethodAs<Int>(
                                "getHeight"
                            )
                        ) < 5

                    if (thumbs == null || !isThumb(current, thumbs[index])) {
                        withContext(Dispatchers.IO) {
                            url =
                                "https://sf1-ttcdn-tos.pstatp.com/obj/${current.callMethod("getUri")}"
                            val req = "okhttp3.Request\$Builder".findClass(cl).new()
                                .callMethod("url", url)?.callMethod("build")
                            val resp = "okhttp3.OkHttpClient".findClass(cl).new()
                                .callMethod("newCall", req)?.callMethod("execute")
                            useCdn = resp?.callMethod("body")?.callMethod("contentType")
                                .toString() != "image/webp"
                        }
                    }
                    val urlList = current.callMethodAs<ArrayList<*>>("getUrlList")
                    val isGif = current.callMethodAs<Boolean>("isGif")
                    if (!useCdn) {
                        url = urlList[0].callMethodAs("getUrl")
                        url = url.substring(0, url.lastIndexOf('.')) + if (isGif) ".gif" else ".png"
                    }
                    urlList[0].callMethod("setUrl", url)
                    callMethod("downloadImage", urlList, isGif)
                }
            }
        }
    }
}